src/codeql.ts

```diff
@@ -230,6 +277,102 @@ async function getSourceLocationPrefix(codeql: string) {
   return resolvedDatabase.sourceLocationPrefix;
 }
 
+interface QueryPackRunResults {
+  queries: Array<{
+    queryPath: string;
+    relativeBqrsFilePath: string;
+    bqrsInfo: BQRSInfo;
+  }>;
+  totalResultsCount: number;
+  resultsBasePath: string;
+}
+
+async function getQueryPackRunResults(
+  codeql: string,
+  databaseName: string,
+  queryPaths: string[],
+  queryPackPath: string,
+  queryPackName: string,
+): Promise<QueryPackRunResults> {
+  // This is where results are saved, according to
+  // https://codeql.github.com/docs/codeql-cli/manual/database-run-queries/
+  const resultsBasePath = `${databaseName}/results`;
+
+  const queries: Array<{
+    queryPath: string;
+    relativeBqrsFilePath: string;
+    bqrsInfo: BQRSInfo;
+  }> = [];
+
+  let totalResultsCount = 0;
+
+  for (const queryPath of queryPaths) {
+    // Calculate the BQRS file path
+    const queryPackRelativePath = path.relative(queryPackPath, queryPath);
+    const parsedQueryPath = path.parse(queryPackRelativePath);
+    const relativeBqrsFilePath = path.join(
+      queryPackName,
+      parsedQueryPath.dir,
+      `${parsedQueryPath.name}.bqrs`,
+    );
+    const bqrsFilePath = path.join(resultsBasePath, relativeBqrsFilePath);
+
+    if (!fs.existsSync(bqrsFilePath)) {
+      throw new Error(
+        `Could not find BQRS file for query ${queryPath} at ${bqrsFilePath}`,
+      );
+    }
+
+    const bqrsInfo = await getBqrsInfo(codeql, bqrsFilePath);
+
+    queries.push({
+      queryPath,
+      relativeBqrsFilePath,
+      bqrsInfo,
+    });
+
+    totalResultsCount += getBqrsResultCount(bqrsInfo);
+  }
+
+  return {
+    totalResultsCount,
+    resultsBasePath,
+    queries,
+  };
+}
+
+async function querySupportsSarif(
+  codeql: string,
+  queryPath: string,
+  bqrsInfo: BQRSInfo,
+): Promise<boolean> {
+  const compatibleQueryKinds = bqrsInfo.compatibleQueryKinds;
```

Comment:
(Open question)

The plumbing of all this is a bit odd. We call `codeql bqrs info` on every BQRS file earlier and store it so we can access it here and also call `codeql resolve metadata`. My brain is just saying it seems a bit off, but I can't think of anything better.

Is there any risk that we'll want to use the metadata somewhere else as well? Should we be storing it in `QueryPackRunResults` or otherwise computing it once and storing it, rather than computing it here?

Reply:
Agree it's a bit off and it's caused me some confusion. However, the changes are following the existing flow. I'd be happy to change it (and I'm glad you're equally finding it odd!), but I think it would be good to do that separately (in another PR/issue).